### PR TITLE
Use GPflow posterior objects to speed up acquisition.

### PR DESCRIPTION
**Pull request by jesnie**
Use GPflow posterior objects to speed up acquisition.
- this still requires new versions of GPflow and GPflux, and maybe the software architecture could be nicer.

### CODE REVIEW

File: trieste/models/interfaces.py [hunk 1]
```diff
@@ -120,6 +120,13 @@ def optimize(self, dataset: Dataset) -> None:
         """
         raise NotImplementedError
 
+    def after_load(self) -> None:
+        """
+        Call this after de-pickling the model, or loading it in another way that bypasses
+        `__init__`.
+        """
+        pass
+
```
Comment (henrymoss) [inline trieste/models/interfaces.py:129]: What does this do?

File: trieste/models/gpflow/models.py [hunk 2]
```diff
@@ -106,6 +115,19 @@ def __init__(
 
         self._ensure_variable_model_data()
 
+        # Cache posterior for fast sequential predictions.
+        # It is important that we create the posterior *after* we ensure that the model data is
+        # variable.
+        self._posterior = model.posterior(PrecomputeCacheType.VARIABLE)
```
Comment (henrymoss) [inline trieste/models/gpflow/models.py:116]: So I didnt have `PrecomputeCacheType.VARIABLE`. Is this why it wasn't updating properly?

File: trieste/models/interfaces.py [hunk 3]
```diff
@@ -120,6 +120,13 @@ def optimize(self, dataset: Dataset) -> None:
         """
         raise NotImplementedError
 
+    def after_load(self) -> None:
+        """
+        Call this after de-pickling the model, or loading it in another way that bypasses
+        `__init__`.
+        """
```
Comment (henrymoss) [inline trieste/models/interfaces.py:127]: Perhaps say why we need this. At the moment its not clear until you look in models.py

File: trieste/models/gpflow/models.py [hunk 4]
```diff
@@ -328,17 +362,17 @@ def conditional_predict_f(
         if isinstance(self.model, SGPR):
             raise NotImplementedError("Conditional predict f is not supported for SGPR.")
 
-        mean_add, cov_add = self.model.predict_f(
-            additional_data.query_points, full_cov=True
+        mean_add, cov_add = self.predict_joint(
+            additional_data.query_points
```
Comment (henrymoss) [inline trieste/models/gpflow/models.py:371]: maybe stress why we dont use the posterior object here (i.e. this is model fitting)

File: trieste/models/gpflow/models.py [hunk 5]
```diff
@@ -632,7 +693,16 @@ def update(self, dataset: Dataset) -> None:
...
+        """
+        Optimize the model with the specified `dataset`.
+
+        :param dataset: The data with which to optimize the `model`.
+        """
+        self.optimizer.optimize(self.model, dataset)
+        self._posterior.update_cache()  # update cached posterior
 
 class VariationalGaussianProcess(
     GPflowPredictor, TrainableProbabilisticModel, SupportsInternalData
```
Comment (henrymoss) [inline trieste/models/gpflow/models.py]: Will you be able to add cache support for VGPs? or is that another PR?

File: tests/unit/models/gpflow/test_models.py [hunk 6]
```diff
@@ -585,6 +587,77 @@ def test_gaussian_process_regression_optimize(compile: bool) -> None:
...
+#     x_predict = np.linspace(0, 5, 2).reshape((-1, 1))
+#     t_0 = time()
+#     [model.predict(x_predict) for _ in range(20)]  # make sequential predictions
+#     time_with_cache = time() - t_0
+#     t_0 = time()
+#     [model.model.predict_f(x_predict) for _ in range(100)]
+#     time_without_cache = time() - t_0
+
+#     npt.assert_array_less(time_with_cache, time_without_cache)
+
```
Comment (henrymoss) [inline tests/unit/models/gpflow/test_models.py]: I think its important to have a test like this one.

**Review by henrymoss — COMMENTED** [review]
Looks promising!

**Comment by hstojic** [comment]
this can now be closed, right?

**Pull request closed without merging**